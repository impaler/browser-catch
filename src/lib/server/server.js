const os = require('os')
const dns = require('dns')
const http = require('http')
const express = require('express')

const DEFAULT_HEADER = {
  'Content-Type': 'text/html',
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Headers': 'Origin, X-Requested-With, Content-Type, Accept'
}

const DEFAULT_ROUTES = [
  {
    name: '/favicon.ico',
    html: '',
    header: {'Content-Type': 'image/x-icon'}
  },
  {
    name: '/',
  },
]

let DEFAULT_ROUTE = {
  headTemplate: context => `<title>${context.name}</title>`,
  htmlTemplate: context => `<html>
  <head>
    ${context.head}
  </head>
  <body>
    ${context.body}
  </body>
</html>`
}

const scriptTemplate = context => `<script>
    ${context.script}
  </script>`

const DEFAULT_OPTIONS = {
  verbose: false,
  routes: []
}

export default function (options) {
  options = Object.assign(DEFAULT_OPTIONS, options) || DEFAULT_OPTIONS
  return new Promise(resolveServer)

  function resolveServer (resolve, reject) {
    let app = express()
    let server = http.createServer(app).listen()
    let port = options.port || server.address().port

    server.close()
      .on('close', () => {
        if (options.verbose) console.log('Server stopped')
        app.listen(port, function () {
          dns.lookup(os.hostname(), onLogHost.bind(this, port))
        })
      })
      .on('error', error => reject(error))

    let routes = [...DEFAULT_ROUTES, ...options.routes]
    for (let route of routes) {
      addRoute(route, app, options)
    }

    DEFAULT_ROUTES[1].body = `<ul>
    ${listRoutes(app)}
</ul>`

    function listRoutes (app) {
      return app._router.stack
        .filter(item => item.name === 'bound dispatch')
        .map(item => `<li><a href="${item.route.path}">${item.route.path}</a></li>`).join('\n')
    }

    function onLogHost (port, error, address) {
      if (error) throw error
      let url = `http://${address}:${port}`
      if (options.verbose) console.log(`Example app listening at ${url}`)

      resolve(Object.assign({},
        {url, address, port, app},
        {add: addRoute.bind(null, app, options)}
      ))
    }
  }
}

function addRoute (route, app, options) {
  if (route.name) {
    app.get(route.name, resolveRoute.bind(null, route, options))
  } else {
    throw new Error('add requires a routePath and content parameter to add a route')
  }
}

function resolveRoute (route, options, request, response) {
  let context = {name: route.name, query: request.query}
  route = Object.assign({}, DEFAULT_ROUTE, route)

  let head = resolveContentSegment('head', route, context)
  let body = resolveContentSegment('body', route, context)
  let htmlContent = resolveContentSegment('html', route,
      Object.assign({}, context, {head, body})
    ) || ''
  let header = Object.assign({}, DEFAULT_HEADER, route.header)

  response.set(header)
  response.send(htmlContent)
}

function resolveContentSegment (key, route, context) {
  let template = route[`${key}Template`]
  if (template) {
    return template(context)
  } else {
    return route[key]
  }
}
