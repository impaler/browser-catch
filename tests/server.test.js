import test from 'ava'
const fetch = require('node-fetch')

import fixtureServer from '../src/lib/fixture-server'
import { HTML_TEMPLATE_FIXTURE } from './fixtures/server/server-tests'
import ERRORS from './fixtures/server/errors'
import FEATURES from './fixtures/server/features'

test('fixture-server to respond with a 404 for unknown content', async t => {
  const serverSettings = await fixtureServer()
  const response = await fetch(`${serverSettings.url}/randomstuff`)

  t.is(response.status, 404)
  t.is(response.statusText, 'Not Found')
})

test('passing query arguments to the context of the htmlTemplate', async t => {
  const ROUTES = {
    routes: [HTML_TEMPLATE_FIXTURE]
  }
  const serverSettings = await fixtureServer(ROUTES)
  const response = await fetch(`${serverSettings.url}/query?apples=green&banana=yellow&square=4&triangle=3`)
  const responseText = await response.text()

  t.snapshot(responseText)
})

test('passing multiple routes to the server', async t => {
  const ROUTES = {
    routes: [...ERRORS, ...FEATURES]
  }
  const serverSettings = await fixtureServer(ROUTES)

  const firstRouteResponse = await fetch(`${serverSettings.url}/cats`)
  const firstRoute = await firstRouteResponse.text()

  const secondRouteResponse = await fetch(`${serverSettings.url}/dogs`)
  const secondRoute = await secondRouteResponse.text()

  const thirdRouteResponse = await fetch(`${serverSettings.url}/error`)
  const thirdRoute = await thirdRouteResponse.text()

  const noRouteResponse = await fetch(`${serverSettings.url}/randomstuff`)
  const noRouteText = {status: noRouteResponse.status, statusText: noRouteResponse.statusText}

  t.snapshot({firstRoute, secondRoute, thirdRoute, noRouteText})
})
