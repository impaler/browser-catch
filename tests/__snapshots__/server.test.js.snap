{
  "passing query arguments to the context of the htmlTemplate": "\n<html>\n  <head>\n    <title>/query</title>\n  </head>\n  <body>\n    <h1>{\"apples\":\"green\",\"banana\":\"yellow\",\"square\":\"4\",\"triangle\":\"3\"}</h1>\n    <h2>Override body</h2>\n  </body>\n</html>",
  "passing multiple routes to the server": {
    "firstRoute": "<html>\n  <head>\n    <title>/cats</title>\n  </head>\n  <body>\n    \n  <ul class=\"breeds\">\n    <li>Siamese</li>\n    <li>Ragdoll</li>\n    <li>Persian</li>\n    <li>Munchkin</li>\n  </ul>\n\n  </body>\n</html>",
    "secondRoute": "<html>\n  <head>\n    <title>/dogs</title>\n  </head>\n  <body>\n    \n  <ul class=\"breeds\">\n    <li>Whippet</li>\n    <li>Staffordshire</li>\n    <li>Grey Hound</li>\n    <li>Boston Terrier</li>\n  </ul>\n  <script type=\"text/javascript\">\n    console.error(\"dogs error\")\n  </script>\n\n  </body>\n</html>",
    "thirdRoute": "<html>\n  <head>\n    <title>/error</title>\n  </head>\n  <body>\n    <script>console.error(\"A regular console.error\")</script>\n  </body>\n</html>",
    "noRouteText": {
      "status": 404,
      "statusText": "Not Found"
    }
  }
}